### ISSUE TITLE
improve the result

### ISSUE THREAD

**Issue by zhaozhongch** [kou1okada/libelas#2]
Thanks for your work but I didn't get desired result from this lib.
I have two pgm images as following
![left_1403637130538319104](https://user-images.githubusercontent.com/25599902/60465177-bc597d00-9c0d-11e9-9b0a-8bb7de82535a.png)
![right_1403637130538319104](https://user-images.githubusercontent.com/25599902/60465186-bfed0400-9c0d-11e9-9886-1812a262e27e.png)
(The upload pictures are png version seems github issue cannot allow me to upload pgm version but in the code I use .pgm picture)
The result is like the follwing
![left_1403637130538319105_disp](https://user-images.githubusercontent.com/25599902/60465420-43a6f080-9c0e-11e9-9bfc-880979e2fc01.png)
![right_1403637130538319104_disp](https://user-images.githubusercontent.com/25599902/60465421-46094a80-9c0e-11e9-9b83-2ed817706c8e.png)
I just use the original demo main.cpp. Anyway to improve it?
Thanks!

**Comment by kou1okada** [comment]
The large part of stereo matching algorithms are assuming the epipolar constraint along the y-axis.
Under this assumption, the same coordinate values of the y-axis on left and right images must ride on same epipolar line.
LIBELAS is also assumed it.
But your images seems to be shifted about 11 pixels along the y-axis.
Also the negative values for disparity are not preferred for many of implementations of stereo matching algorithms.
For more details, read original thesis and wikipedia.
* http://www.cvlibs.net/publications/Geiger2010ACCV.pdf
* https://en.wikipedia.org/wiki/Epipolar_geometry

By the way, for locations closed to the camera, I could not be matched  images by binocular stereopsis even with my bare eyes. 
Thus, it should be not easy to match them by the stereo matching algorithms.

My result is following:
```
# Download images
wget https://user-images.githubusercontent.com/25599902/60465177-bc597d00-9c0d-11e9-9b0a-8bb7de82535a.png -O left.png
wget https://user-images.githubusercontent.com/25599902/60465186-bfed0400-9c0d-11e9-9886-1812a262e27e.png -O right.png
# Translate the right image by (-15,-10) with convert command of ImageMagick.
convert right.png -distort Affine '0,0 -15,-11' right-15-11.png
# Convert source images from png to pgm.
for i in {left,right}*.png; do convert -verbose $i ${i%.*}.pgm; done
# Get disparity maps with LIBELAS.
./elas.exe left.pgm right-15-11.pgm
# Convert result images from pgm to png. 
for i in {left,right}*_disp.pgm; do convert -verbose $i ${i%.*}.png; done
```
left.png
![left](https://user-images.githubusercontent.com/1631846/60479380-9ccf5c80-9cc0-11e9-82e5-5cc4523bacd9.png)
right-15-11.png
![right-15-11](https://user-images.githubusercontent.com/1631846/60479408-b07ac300-9cc0-11e9-8ea2-dac6c13cfcf8.png)
left_disp.png
![left_disp](https://user-images.githubusercontent.com/1631846/60479452-def89e00-9cc0-11e9-80a8-80729296db6d.png)
right-15-11_disp.png
![right-15-11_disp](https://user-images.githubusercontent.com/1631846/60479464-e61fac00-9cc0-11e9-8984-155ca31a91f2.png)

**Comment by zhaozhongch** [comment]
Thanks for your reply. I'll try to play with the parameter for a while

**Comment by kou1okada** [comment]
Wait, sorry.
May be my post, which was already removed, is not correct.

**Comment by kou1okada** [comment]
The cause of the problem was not only translation but also rotation.

**Comment by kou1okada** [comment]
The larger part of stereo matching algorithms are assuming the epipolar constraint along the y-axis.
Under this assumption, the same coordinate values of the y-axis on left and right images must ride on same epipolar line.
LIBELAS is also assumed it.
But your images seem to be shifted about 11 pixels along the y-axis.
Also the negative values for the disparity are not preferred for many of the implementations of stereo matching algorithms.
For more details, read original thesis and Wikipedia.
* http://www.cvlibs.net/publications/Geiger2010ACCV.pdf
* https://en.wikipedia.org/wiki/Epipolar_geometry

By the way, for locations closed to the camera, I could not be matched  images by binocular stereopsis even with my bare eyes. 
Thus, it should be not easy to match them by the stereo matching algorithms.

My result is following:
```
# Download images
wget https://user-images.githubusercontent.com/25599902/60465177-bc597d00-9c0d-11e9-9b0a-8bb7de82535a.png -O left.png
wget https://user-images.githubusercontent.com/25599902/60465186-bfed0400-9c0d-11e9-9886-1812a262e27e.png -O right.png
# Translate the right image by (-15,-11) with convert command of ImageMagick.
convert right.png -distort Affine '0,0 -15,-11' right-15-11.png
# Convert source images from png to pgm.
for i in {left,right}*.png; do convert -verbose $i ${i%.*}.pgm; done
# Get disparity maps with LIBELAS.
./elas.exe left.pgm right-15-11.pgm
# Convert result images from pgm to png. 
for i in {left,right}*_disp.pgm; do convert -verbose $i ${i%.*}.png; done
```
left.png
![left](https://user-images.githubusercontent.com/1631846/60479380-9ccf5c80-9cc0-11e9-82e5-5cc4523bacd9.png)
right-15-11.png
![right-15-11](https://user-images.githubusercontent.com/1631846/60479408-b07ac300-9cc0-11e9-8ea2-dac6c13cfcf8.png)
left_disp.png
![left_disp](https://user-images.githubusercontent.com/1631846/60479452-def89e00-9cc0-11e9-80a8-80729296db6d.png)
right-15-11_disp.png
![right-15-11_disp](https://user-images.githubusercontent.com/1631846/60479464-e61fac00-9cc0-11e9-8984-155ca31a91f2.png)

**Comment by kou1okada** [comment]
The larger part of stereo matching algorithms are assuming the epipolar constraint along the y-axis.
Under this assumption, the same coordinate values of the y-axis on left and right images must ride on same epipolar line.
LIBELAS is also assumed it.
But your images seem to have translation and rotation.
Also the negative values for the disparity are not preferred for many of the implementations of stereo matching algorithms.
For more details, read original thesis and Wikipedia.
* http://www.cvlibs.net/publications/Geiger2010ACCV.pdf
* https://en.wikipedia.org/wiki/Epipolar_geometry

By the way, for locations closed to the camera, I could not be matched  images by binocular stereopsis even with my bare eyes. 
Thus, it should be not easy to match them by the stereo matching algorithms.

My result is following:
```bash
# Download images
wget https://user-images.githubusercontent.com/25599902/60465177-bc597d00-9c0d-11e9-9b0a-8bb7de82535a.png -O left.png
wget https://user-images.githubusercontent.com/25599902/60465186-bfed0400-9c0d-11e9-9886-1812a262e27e.png -O right.png
# Fix rotation and parallel translation for the right image with convert command of ImageMagick.
convert -verbose right.png -rotate 0.35 -crop 752x480+0+0 -distort Affine '0,0 -12,-14' right-fixed.png
# Convert source images from png to pgm.
for i in {left,right}*.png; do convert -verbose $i ${i%.*}.pgm; done
# Get disparity maps with LIBELAS.
./elas left.pgm right-fixed.pgm
# Convert result images from pgm to png. 
for i in {left,right}*_disp.pgm; do convert -verbose $i ${i%.*}.png; done
```
left.png
![left](https://user-images.githubusercontent.com/1631846/60479380-9ccf5c80-9cc0-11e9-82e5-5cc4523bacd9.png)
right-fixed.png
![right-fixed](https://user-images.githubusercontent.com/1631846/60510802-52270200-9d0b-11e9-8d4e-bc1a88658dbb.png)
left_disp.png
![left_disp](https://user-images.githubusercontent.com/1631846/60510823-5ce19700-9d0b-11e9-90a5-9764da9e00ec.png)
right-fixed_disp.png
![right-fixed_disp](https://user-images.githubusercontent.com/1631846/60510830-61a64b00-9d0b-11e9-856e-153609ae617c.png)

**Comment by kou1okada** [comment]
Hmm,,,
It's so sensitive to deviations from the epipolar constraint.

**Comment by zhaozhongch** [comment]
Hi thank you very much for solving this.
My data comes from a dataset called euroc dataset 
https://projects.asl.ethz.ch/datasets/doku.php?id=kmavvisualinertialdatasets
This dataset has image in `cam0` folder and `cam1` folder if you download the dataset ASL format, which corresponding to the left_camera and right_camera. I download the `Machine Hall 3` one. I just pick one image in the cam0 folder and then pick another image with the same timestamp in the cam1 folder.
considering you say the picture is too close to the camera. I choose another set of picture from the`Machine Hall 3`. Which is the following
![left_1403637159038319104](https://user-images.githubusercontent.com/25599902/60544607-5dac0600-9cd6-11e9-8392-2803c30f18f8.png)
![right_1403637159038319104](https://user-images.githubusercontent.com/25599902/60544612-600e6000-9cd6-11e9-84c4-414e6e545552.png)
The result using the elas is 
![elas](https://user-images.githubusercontent.com/25599902/60545570-9ea51a00-9cd8-11e9-8343-e9f59c62925b.png)
Considering the first example this looks better.
I also tried to use opencv's build-in way to compute the disparity map, the code comes from 
https://docs.opencv.org/3.1.0/d3/d14/tutorial_ximgproc_disparity_filtering.html
The result is 
![sgbm_wls_onfig](https://user-images.githubusercontent.com/25599902/60545741-00fe1a80-9cd9-11e9-9363-8efa2c65d3ac.png)
From this two view seems the result from opencv is slightly better. The important thing is that using libelas it costs about 200ms the get the disparity map but the opencv one costs 20 to 25 ms.
This is my result so far, I'll again adjust the parameter to see if I can get better result
Thanks!

**Comment by kou1okada** [comment]
This should be improved the result by calibration, too.
It seems better conditions than first case.

**Comment by oleg-alexandrov** [comment]
I played quite a bit with both LIBELAS and OpenCV's SGBM algorithm. LIBELAS does well, but the way it takes its inputs and outputs is a bit awkward. The biggest limitation I see is the constraint that in LIBEAS the disparity must be positive. I solve this by artificially padding the left image on the left and the right image on the right with zeros with a value big enough that LIBELAS does compute a positive disparity, then I undo this in the disparity returned from LIBELAS.

Also, LIBELAS does not handle very gracefully the case when the input data has missing values, say a certain portion of the scene is seen only in one image and I fill with zeros the corresponding area for the second image. It can return a junk disparity, but which can be filtered out as its values have a jump compared to regular disparities. 

Anyhow, I made a fork of this repo and tweaked the main program to deal with these issues. I also made this program take the parameters LIBELAS uses later as command-line arguments, and read and write its images as TIF with float values rather than using 8-bit images on input and output.

My fork is at https://github.com/NeoGeographyToolkit/libelas
